### PR TITLE
Add upload writer behavior for "stream to stream" uploads

### PR DESCRIPTION
**Pull request by chrismccord**

### CODE REVIEW

File: lib/phoenix_live_view/upload_channel.ex [hunk 1]
```diff
@@ -154,25 +183,44 @@ defmodule Phoenix.LiveView.UploadChannel do
...
+        {:ok, socket} -> {:ok, assign(socket, done?: true)}
+        {:error, reason, new_socket} -> {:error, reason, new_socket}
+      end
     else
-      socket
+      {:ok, socket}
     end
   end
 
   defp close_file(socket) do
```
Comment (josevalim) [inline lib/phoenix_live_view/upload_channel.ex:210]: Rename to `close_writer`?

File: lib/phoenix_live_view/upload_config.ex [hunk 2]
```diff
@@ -270,6 +273,26 @@ defmodule Phoenix.LiveView.UploadConfig do
...
+          Only a function which accepts the socket is supported. Got:
+
+          #{inspect(other)}
+          """
+
+        :error ->
+          fn _socket -> {Phoenix.LiveView.UploadWriter, []} end
+      end
+
+
```
Comment (josevalim) [inline lib/phoenix_live_view/upload_config.ex:294]:
```suggestion
```

File: lib/phoenix_live_view/upload_writer.ex [hunk 3]
```diff
@@ -0,0 +1,32 @@
+defmodule Phoenix.LiveView.UploadWriter do
+
```
Comment (josevalim) [inline lib/phoenix_live_view/upload_writer.ex:2]: We need docs and I would move the default implementation to a separate file. :)

File: lib/phoenix_live_view/upload_config.ex [hunk 4]
```diff
@@ -270,6 +273,26 @@ defmodule Phoenix.LiveView.UploadConfig do
           nil
       end
 
+    writer =
+      case Keyword.fetch(opts, :writer) do
+        {:ok, func} when is_function(func) ->
+          func
+
+        {:ok, other} ->
+          raise ArgumentError, """
```
Comment (josevalim) [inline lib/phoenix_live_view/upload_config.ex:282]: I am a bit confused. Here we say it has to be a function but in the writer we say it is either a module or a function. Which one is it?

Comment (chrismccord) [reply lib/phoenix_live_view/upload_config.ex:282]: it is both. I will update!

File: assets/js/phoenix_live_view/entry_uploader.js [hunk 5]
```diff
@@ -9,12 +9,14 @@ export default class EntryUploader {
     this.offset = 0
     this.chunkSize = chunkSize
     this.chunkTimer = null
+    this.errored = false
     this.uploadChannel = liveSocket.channel(`lvu:${entry.ref}`, {token: entry.metadata()})
   }
 
   error(reason){
+    if(this.errored){ return }
```
Comment (jeregrine) [inline assets/js/phoenix_live_view/entry_uploader.js:17]: should we console.error the error every time? just in case?

**Review by jeregrine — COMMENTED** [review]
Agree with @josevalim on Docs and one JS comment

**Review by mcrumm — COMMENTED** [review]
This is spectacular, and it works almost exactly as expected 🥳 

One small suggestion– allow the writer access to the Entry on `init` so that metadata can be retrieved. Rather than try to make the required suggestions here I opened a PR:

* #2723

File: lib/phoenix_live_view/upload_config.ex (hunk 4 above)
Comment (mcrumm) [reply lib/phoenix_live_view/upload_config.ex:283]: we only accept a 3-arity function now :)

File: lib/phoenix_live_view/upload_writer.ex [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  @callback init(opts :: term) :: {:ok, state :: term} | {:error, term}
+  @callback meta(state :: term) :: map
+  @callback write_chunk(state :: term, data :: binary) :: {:ok, state :: term} | {:error, term}
```
Comment (mcrumm) [inline lib/phoenix_live_view/upload_writer.ex:5]: Two questions, both pedantic:

1. Should the order of the arguments be flipped? i.e. `(data :: binary, state :: term)`?
2. Should we rename to `handle_chunk`?

```suggestion
  @callback handle_chunk(data :: binary, state :: term) :: {:ok, state :: term} | {:error, term}
```

Comment (chrismccord) [reply lib/phoenix_live_view/upload_writer.ex:5]: definitely flip the arguments. Naming not sure. It is called a "writer", so I am not sure about handle_chunk, but now I am having an existential crisis about it :D

Comment (mcrumm) [reply lib/phoenix_live_view/upload_writer.ex:5]: `write_chunk` is fine, no existential crisis necessary! 🤣 i'll update the suggestion

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**